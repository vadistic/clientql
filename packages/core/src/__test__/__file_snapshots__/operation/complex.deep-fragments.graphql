query FindEventsAtVenueMyQuery($venueId: ID!) {
  findEventsAtVenue(venueId: $venueId) {
    ...EventDeep
  }
}

fragment EventDeep on Event {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  ... on Concert {
    ...ConcertDeep
  }
  ... on Festival {
    ...FestivalDeep
  }
  ... on Conference {
    ...ConferenceDeep
  }
}

fragment VenueDeep on Venue {
  __typename
  id
  name
  address
  maxOccupancy
}

fragment ConcertDeep on Concert {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  previousVenues {
    ...VenueDeep
  }
  minAgeRestriction
  performingBand {
    ...PerformerDeep
  }
}

fragment PerformerDeep on Performer {
  __typename
  id
  name
  contact
  speciality
  fee
}

fragment FestivalDeep on Festival {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  performers {
    ...PerformerDeep
  }
}

fragment ConferenceDeep on Conference {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  speakers {
    ...SpeakerDeep
  }
  workshops
}

fragment SpeakerDeep on Speaker {
  __typename
  id
  name
  contact
}
