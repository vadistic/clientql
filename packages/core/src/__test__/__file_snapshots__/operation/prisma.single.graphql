query UserQuery($where: UserWhereUniqueInput!) {
  user(where: $where) {
    ...UserFlatPartial
    posts {
      ...PostFlatPartial
    }
  }
}

fragment UserFlatPartial on User {
  __typename
  id
  nationality
  email
  name
}

fragment PostFlatPartial on Post {
  __typename
  id
  content
  createdAt
  published
  title
  updatedAt
}
