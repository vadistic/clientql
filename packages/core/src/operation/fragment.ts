import { DocumentNode } from 'graphql'
import { createFragment, FragmentName, Typename, wrapDocument } from '../ast'
import { CoreProps } from '../core'
import { buildSelections } from './selections'
import { FragmentResult, SelectionsResult } from './types'
import { retriveFragmentsFromCache } from './utils'

/**
 * fragments is just wrapped array of selections on some type,
 * but let's have helper for it
 *
 * fragment with type NONE will be a fragment,
 * but will not use any fragment helpers
 */

export const buildFragment = (props: CoreProps) => (
  typename: Typename,
  name?: FragmentName,
): FragmentResult | undefined => {
  const { selections, complete, flat, fragmentNames } = buildSelections(props)(
    typename,
  )

  // invalid typename or empty with current config
  if (!selections) {
    return undefined
  }

  // need to namespace non-complete fragments,
  // user can produce them but thay should be named differently
  // I would not like to produce 2 different fragments with same name
  // TODO: think how to name it
  const fragmentName = getFragmentName(props)({ typename, flat, complete })

  // shortcircuit on cached result
  if (!name && props.cache.fragments.has(fragmentName)) {
    return props.cache.fragments.get(fragmentName)!
  }

  const fragment = createFragment({
    condition: typename,
    fragmentname: name || fragmentName,
    selections,
  })

  return {
    fragment,
    fragmentNames,
    complete,
    flat,
  }
}

/**
 * same as build operation doc, create fragment by typename
 */
export const buildFragmentDoc = (props: CoreProps) => (
  typename: Typename,
  name?: FragmentName,
): DocumentNode | undefined => {
  const fragmentResult = buildFragment(props)(typename, name)

  if (!fragmentResult) {
    return
  }

  const { fragmentNames, fragment } = fragmentResult

  const fragments = retriveFragmentsFromCache(props)(fragmentNames)

  return wrapDocument(fragment, ...fragments)
}

/**
 * helper to normalise fragment naming
 */
export interface GetFragmentNameOptions {
  typename: Typename
  complete: boolean
  flat: boolean
}

export const getFragmentName = (props: CoreProps) => ({
  complete,
  flat,
  typename,
}: GetFragmentNameOptions): FragmentName => {
  if (flat) {
    return typename + props.config.fragmentFlatSuffix
  }

  if (complete) {
    return typename + props.config.fragmentDeepSuffix
  }

  // deep, non-complete selection is a (deep) partial
  return typename + props.config.fragmentPartialSuffix
}
