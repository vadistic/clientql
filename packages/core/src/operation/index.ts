export * from './selection'
export * from './operation'
