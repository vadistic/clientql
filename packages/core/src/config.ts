import { DocumentNode, FragmentDefinitionNode } from 'graphql'
import { SelectionResult } from './operation'
import { createTypeGraph, TypeGraph } from './type-graph'

export enum FragmentType {
  /** use flat fragments when possible */
  FLAT = 'Flat',
  /** use one deep fragment */
  DEEP = 'Deep',
  /** do not use fragments */
  NONE = 'None',
}

/**
 * Config for runtime client
 */

export interface CoreConfig {
  /**
   * how deep operations/fragments should query
   * 0 === flat
   * @default 3
   */
  maxDepth: number

  /**
   * idea is that operation can query using different kind of fragments
   */
  useFragments: FragmentType

  addTypename: boolean
}

export const defaultCoreConfig: CoreConfig = {
  maxDepth: 4,
  useFragments: FragmentType.FLAT,
  addTypename: true,
}

/**
 * Uniform props for utils options
 */
export interface CoreProps {
  config: CoreConfig
  graph: TypeGraph
  // lazy map for reusable fragments
  fragments: Map<string, FragmentDefinitionNode>
  // testing to avoid traversing whole tree few times each op
  // lets store only complete ones (should cover 90% of the tree)
  selections: Map<string, SelectionResult>
}

export const getCoreProps = (
  doc: DocumentNode,
  config?: Partial<CoreConfig>,
): CoreProps => ({
  graph: createTypeGraph(doc),
  config: {
    ...defaultCoreConfig,
    ...config,
  },
  fragments: new Map(),
  selections: new Map(),
})
