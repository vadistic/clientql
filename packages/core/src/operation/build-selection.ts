import { Kind, SelectionNode } from 'graphql'
import {
  createField,
  createFragment,
  createFragmentSpread,
  createInlineFragment,
  Fieldname,
  FragmentName,
  isEnumTypeDefinitionNode,
  isInterfaceTypeDefinitonNode,
  isObjectTypeDefinitionNode,
  isScalarTypeDefinitionNode,
  isUnionTypeDefinitionNode,
  Typename,
} from '../ast'
import { FragmentType } from '../config'
import { CoreProps } from '../core'
import { Edge } from '../graph'
import { isNotEmpty } from '../utils'
import {
  FragmentResult,
  NestableSelectionResult,
  SelectionsResult,
} from './types'
import {
  isStackCircural,
  lastFieldname,
  lastTypename,
  onlyUnique,
  replaceLastTypename,
} from './utils'

/**
 * wraper with friendlier api for building selection of some `__typename`
 *
 * returns noop on invalid typename
 */
export const buildSelections = (props: CoreProps) => (typename: Typename) => {
  if (!props.graph.has(typename)) {
    return noopResult
  }

  return buildRecursiveSelections(props)([['$ROOT', typename]])
}

/**
 * empty result
 */
const noopResult: SelectionsResult = {
  complete: true,
  flat: true,
  dependencies: [],
  selections: undefined,
}

/**
 * build selection based on stack
 *
 * main reducer:
 * - validate
 * - build flat selection
 * - build/retrive cached nested selections
 */
const buildRecursiveSelections = (props: CoreProps) => (
  stack: Edge[],
): SelectionsResult => {
  const typename = lastTypename(stack)
  const fieldname = lastFieldname(stack)

  const vtx = props.graph.get(typename)
  const allowedDepth = props.config.maxDepth - (stack.length - 1)

  // trivial flat cases, always allowed
  if (
    // scalar not in schema or explicit scalar
    !vtx ||
    isEnumTypeDefinitionNode(vtx.value) ||
    isScalarTypeDefinitionNode(vtx.value)
  ) {
    return {
      ...noopResult,
      selections: [createField({ fieldname })],
    }
  }

  // disable nesting when depth 0 and circural queries
  if (allowedDepth < 1 || isStackCircural(stack)) {
    return { ...noopResult, complete: false }
  }

  // load from cache or build
  const nestedSelection = props.cache.selections.has(typename)
    ? props.cache.selections.get(typename)!
    : buildNestedSelections(props)(stack)

  // another noop
  if (!isNotEmpty(nestedSelection.selections)) {
    return noopResult
  }

  return {
    ...nestedSelection,
    // mark as not flat here
    flat: false,
  }
}

/**
 * wraps non-flat selection result in field and passthrough flat fields
 *
 * I want `buildSelections` fn to retun flat/spreaded result for easier use in other modules,
 * but recursion needs to wrap nested in - well nested field
 * hence this helper
 */

const wrapNestedSelections = (
  selectionsResult: SelectionsResult,
  fieldname: Fieldname,
): NestableSelectionResult => {
  if (!isNotEmpty(selectionsResult.selections)) {
    return noopResult
  }

  if (selectionsResult.flat) {
    return {
      ...selectionsResult,
      selection: selectionsResult.selections[0],
    }
  }

  return {
    ...selectionsResult,
    selection: createField({
      fieldname,
      selections: selectionsResult.selections,
    }),
  }
}

/**
 * build selection for nested types
 */
const buildNestedSelections = (props: CoreProps) => (
  stack: Edge[],
): SelectionsResult => {
  const typename = lastTypename(stack)
  const vtx = props.graph.get(typename)!

  let complete = true
  let flat = true

  const selections: SelectionNode[] = []
  const dependencies: FragmentName[] = []

  /*
   * handle ObjectTypes
   */

  if (isObjectTypeDefinitionNode(vtx.value)) {
    for (const edge of vtx.edgesArr || []) {
      const fieldname = edge[0]
      const recursive = buildRecursiveSelections(props)([...stack, edge])
      const wrapped = wrapNestedSelections(recursive, fieldname)

      if (!wrapped.selection) {
        complete = false
        continue
      }

      selections.push(wrapped.selection)
      dependencies.push(...wrapped.dependencies)

      // basically AND gate^^
      flat = flat && recursive.flat
      complete = complete && recursive.complete
    }
  }

  /*
   * handle UnionTypes
   */

  if (isUnionTypeDefinitionNode(vtx.value)) {
    // weird case of union without members
    if (!vtx.implementations) {
      return noopResult
    }

    for (const implem of vtx.implementations) {
      // replacing last stack entry when resolving union
      // not wrapping result since inlines should be on the same level as the rest
      const recursive = buildRecursiveSelections(props)(
        replaceLastTypename(stack, implem),
      )

      if (!isNotEmpty(recursive.selections)) {
        complete = complete && recursive.complete
        continue
      }

      // but wrapping in inline fragment
      selections.push(createInlineFragment(implem, recursive.selections))
      dependencies.push(...recursive.dependencies)

      flat = flat && recursive.flat
      complete = complete && recursive.complete
    }
  }

  /*
   * handle InterfaceTypes
   */

  if (isInterfaceTypeDefinitonNode(vtx.value)) {
    // handle own fields
    for (const edge of vtx.edgesArr || []) {
      const fieldname = edge[0]
      const recursive = buildRecursiveSelections(props)([...stack, edge])
      const wrapped = wrapNestedSelections(recursive, fieldname)

      if (!wrapped.selection) {
        complete = false
        continue
      }

      selections.push(wrapped.selection)
      dependencies.push(...wrapped.dependencies)

      flat = flat && recursive.flat
      complete = complete && recursive.complete
    }

    // handle possible implementations
    for (const implem of vtx.implementations || []) {
      const recursive = buildRecursiveSelections(props)(
        replaceLastTypename(stack, implem),
      )

      if (!isNotEmpty(recursive.selections)) {
        complete = complete && recursive.complete
        continue
      }

      selections.push(createInlineFragment(implem, recursive.selections))
      dependencies.push(...recursive.dependencies)

      flat = flat && recursive.flat
      complete = complete && recursive.complete
    }
  }

  /*
   * handle fragmenting and caching
   */

  return cacheAndFragment(props)(stack, {
    complete,
    flat,
    dependencies,
    selections,
  })
}

const cacheAndFragment = (props: CoreProps) => (
  stack: Edge[],
  result: SelectionsResult,
): SelectionsResult => {
  if (!isNotEmpty(result.selections)) {
    return result
  }

  const typename = lastTypename(stack)

  // ensure unique fragmentnames from nested deps
  const uniqueDependencies = onlyUnique(result.dependencies)
  let fragmented = { ...result, dependencies: uniqueDependencies }

  /*
   * selecions with DEEP fragments are fragmented only when selection was complete
   */

  if (props.config.useFragments === FragmentType.DEEP && result.complete) {
    const fragmentname: FragmentName =
      typename + (result.flat ? 'Flat' : 'Deep')

    if (!props.cache.fragments.has(fragmentname)) {
      const fragmentResult: FragmentResult = {
        fragment: createFragment({
          condition: typename,
          fragmentname,
          selections: result.selections,
        }),
        dependencies: uniqueDependencies,
        flat: result.flat,
        complete: true,
      }

      props.cache.fragments.set(fragmentname, fragmentResult)
    }

    fragmented = {
      ...fragmented,
      selections: [createFragmentSpread(fragmentname)],
      dependencies: [fragmentname, ...uniqueDependencies],
    }
  }

  /*
   * selecions with FLAT can always store flat part of selection
   * (inlines & spreads are not considered flat part)
   */

  if (props.config.useFragments === FragmentType.FLAT) {
    const fragmentname: FragmentName = typename + 'Flat'

    const { flat: flatSelections, deep: deepSelections } = result.flat
      ? { flat: result.selections, deep: [] }
      : result.selections.reduce(
          (acc, node) => {
            if (node.kind === Kind.FIELD && !node.selectionSet) {
              acc.flat.push(node)
            } else {
              acc.deep.push(node)
            }

            return acc
          },
          {
            flat: [] as SelectionNode[],
            deep: [] as SelectionNode[],
          },
        )

    if (!props.cache.fragments.has(fragmentname)) {
      const fragmentResult: FragmentResult = {
        fragment: createFragment({
          condition: typename,
          fragmentname,
          selections: flatSelections,
        }),
        // flat fragment has no dependencies
        dependencies: [],
        flat: true,
        complete: true,
      }

      props.cache.fragments.set(fragmentname, fragmentResult)
    }

    fragmented = {
      ...fragmented,
      selections: [createFragmentSpread(fragmentname), ...deepSelections],
      dependencies: [fragmentname, ...uniqueDependencies],
    }
  }

  /*
   * write to cache,
   * if selection is complete (complex and uneffective to store partial ones)
   */
  if (fragmented.complete && !props.cache.selections.has(typename)) {
    props.cache.selections.set(typename, fragmented)
  }

  // return
  return fragmented
}
