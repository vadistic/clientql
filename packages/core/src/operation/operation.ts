import {
  ArgumentNode,
  FieldDefinitionNode,
  FieldNode,
  FragmentDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  VariableDefinitionNode,
} from 'graphql'
import {
  createField,
  createOperation,
  createOperationArgument,
  createOperationVariable,
  Fieldname,
  Typename,
  wrapDocument,
} from '../ast'
import { CoreProps } from '../config'
import { buildSelection, retriveCacheFragments } from './selection'

/**
 * this needs to build operation from arr of fieldnames
 */

export const buildOperation = (props: CoreProps) => (path: Fieldname[]) => {
  const [head, ...tail] = path
  const { type, vtx } = determineOperationType(props)(head)

  // head not on any of roots
  if (!vtx || !type) {
    return
  }

  const { fragments, selections, variables } = buildDeepOperation(props)(
    vtx.name,
    path,
  )

  const operation = createOperation({
    name: 'TODO',
    selections,
    type,
    variables,
  })

  return wrapDocument(operation, ...fragments)
}

interface BuildDeepOperation {
  variables: VariableDefinitionNode[]
  fragments: FragmentDefinitionNode[]
  selections: SelectionNode[]
}

const buildDeepOperation = (props: CoreProps) => (
  target: Typename,
  path: Fieldname[],
  level = 0,
): BuildDeepOperation => {
  const [head, ...tail] = path

  const vtx = props.graph.get(target)

  if (!vtx) {
    throw Error('Invalid operation target')
  }

  const edgeData = vtx.weightsMap!.get(head)
  const nextTarget = vtx.edgesMap!.get(head)

  if (!edgeData || !nextTarget) {
    throw Error('Invalid operation path')
  }

  const argSuffix = level > 0 ? '' + level : ''

  const { variables, args } = (edgeData.arguments || []).reduce(
    (acc, input) => {
      acc.args.push(createOperationArgument(argSuffix)(input))
      acc.variables.push(createOperationVariable(argSuffix)(input))
      return acc
    },
    {
      variables: [] as VariableDefinitionNode[],
      args: [] as ArgumentNode[],
    },
  )

  // base
  if (tail.length === 0) {
    const { selections, fragmentnames } = buildSelection(props)([
      [head, nextTarget],
    ])

    const selection = {
      // this should always be field on the top level
      ...(selections[0] as FieldNode),
      // inject args
      arguments: args,
    }

    return {
      fragments: retriveCacheFragments(props)(fragmentnames),
      variables,
      selections: [selection],
    }
  } else {
    // not a base
    const {
      selections,
      fragments,
      variables: deepVariables,
    } = buildDeepOperation(props)(nextTarget, tail, level + 1)

    const nextSelection: FieldNode = createField({
      fieldname: head,
      arguments: args,
      selections,
    })

    return {
      variables: [...variables, ...deepVariables],
      fragments,
      selections: [nextSelection],
    }
  }
}

export const determineOperationType = (props: CoreProps) => (
  head: Fieldname,
) => {
  const query = props.graph.get('Query')
  if (query && query.edgesMap && query.edgesMap.has(head)) {
    return { vtx: query, type: 'query' as OperationTypeNode }
  }

  const mutation = props.graph.get('Mutation')
  if (mutation && mutation.edgesMap && mutation.edgesMap.has(head)) {
    return { vtx: query, type: 'mutation' as OperationTypeNode }
  }

  const subscription = props.graph.get('subscription')
  if (
    subscription &&
    subscription.edgesMap &&
    subscription.edgesMap.has(head)
  ) {
    return { vtx: query, type: 'subscription' as OperationTypeNode }
  }

  return { vtx: undefined, type: undefined }
}
