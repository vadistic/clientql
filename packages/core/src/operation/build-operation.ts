import {
  ArgumentNode,
  DocumentNode,
  FieldNode,
  InlineFragmentNode,
  OperationTypeNode,
  VariableDefinitionNode,
} from 'graphql'
import {
  createField,
  createInlineFragment,
  createOperation,
  createOperationArgument,
  createOperationVariable,
  Typename,
  wrapDocument,
} from '../ast'
import { CoreProps } from '../config'
import { capitalise } from '../utils'
import { buildSelections } from './build-selection'
import { operationTypeToRootTypename } from './root'
import {
  OperationEdge,
  OperationResult,
  OperationSelectionResult,
} from './types'
import { retriveFragmentsFromCache } from './utils'

/**
 * operation builder transform OperationEdges/paths to:
 *  - empty selection(typename only) + variables
 *  - and then use `buildSelection` on the path end
 *
 * OperationEdge tuple represents [fieldname, typename]
 *
 * - fieldname on ObjectTypes or InterfaceType own fields
 * - typename on UnionType/InterfaceType inlines
 *
 * (this seems like a logic way to support entering inlines)
 *
 * ! first fieldname should be OperationType/ Root
 */

export const buildOperation = (props: CoreProps) => (
  path: OperationEdge[],
): OperationResult => {
  const [head] = path
  const operationType = head[0] as OperationTypeNode | undefined

  if (!operationType) {
    throw Error(`buildOperation: operationType is missing in path`)
  }

  const operationTypes = ['query', 'mutation', 'subscription']

  if (!operationTypes.includes(operationType)) {
    throw Error(`buildOperation: invalid operationType (${operationType})`)
  }

  const rootTypename = operationTypeToRootTypename(props.graph)(operationType)

  if (!rootTypename) {
    throw Error(`buildOperation: there is no schema node for ${operationType}`)
  }

  /**
   * Joining all operation field + Inline Typename selections + rootTypename suffix
   * so ['user','friends','WorkFirend','name'] =>
   * UsersFriendsWorkFriendNameQuery
   */
  const operationName =
    path
      .slice(1)
      .map(([fieldname, typename]) => capitalise(fieldname || typename || ''))
      .join('') + capitalise(rootTypename)

  /**
   * use caching
   */
  if (props.operations.has(operationName)) {
    return props.operations.get(operationName)!
  }

  const { selections, variables, ...rest } = buildOperationSelection(props)(
    rootTypename,
    // without first segment since parent root is resolved
    path.slice(1),
  )

  const operation = createOperation({
    name: operationName,
    selections,
    type: operationType,
    variables,
  })

  const operationResult = {
    ...rest,
    operation,
  }

  // wrtie to cache
  props.operations.set(operationName, operationResult)

  return operationResult
}

/**
 * helper to build operation + retrive fragments
 *
 */
export const buildOperationDoc = (props: CoreProps) => (
  path: OperationEdge[],
): DocumentNode => {
  const { dependencies, operation } = buildOperation(props)(path)
  const fragments = retriveFragmentsFromCache(props)(dependencies)

  return wrapDocument(operation, ...fragments)
}

/**
 * recursively build operation selection and then base selection
 *
 * ! path without first segment since parent is provided
 */
const buildOperationSelection = (props: CoreProps) => (
  parent: Typename,
  path: OperationEdge[],
  level = 0,
): OperationSelectionResult => {
  const [[fieldname, typename], ...tail] = path

  /*
   * entering union/interface inline, level stay the same
   */
  if (typename) {
    const { selections, ...rest } = buildOperationSelection(props)(
      typename,
      tail,
      level,
    )

    const inlineSelection: InlineFragmentNode = createInlineFragment(
      typename,
      selections,
    )

    return {
      ...rest,
      selections: [inlineSelection],
    }
  }

  /*
   * validate path/target
   */
  const vtx = props.graph.get(parent)

  if (!vtx || !vtx.weightsMap || !vtx.edgesMap) {
    throw Error('Invalid operation path')
  }

  if (!fieldname) {
    throw Error('Invalid operation path')
  }

  const edgeData = vtx.weightsMap.get(fieldname)
  const child = vtx.edgesMap.get(fieldname)

  if (!edgeData || !child || !fieldname) {
    throw Error('Invalid operation path')
  }

  /*
   * build this level's variables/args
   * top level variables are without suffixes, nested get nesting level
   */
  const argSuffix = level > 0 ? '' + level : ''

  const inputs = (edgeData.arguments || []).reduce(
    (acc, input) => {
      acc.args.push(createOperationArgument(argSuffix)(input))
      acc.vars.push(createOperationVariable(argSuffix)(input))
      return acc
    },
    {
      vars: [] as VariableDefinitionNode[],
      args: [] as ArgumentNode[],
    },
  )

  /*
   * build nested operation
   */
  if (tail.length > 0) {
    const { selections, variables, ...rest } = buildOperationSelection(props)(
      child,
      tail,
      level + 1,
    )

    // TODO: add typename here?

    const deepSelection: FieldNode = createField({
      fieldname,
      arguments: inputs.args,
      selections,
    })

    return {
      ...rest,
      variables: [...inputs.vars, ...variables],
      selections: [deepSelection],
    }
  }

  /*
   * base selection, no nested variables
   */
  if (tail.length === 0) {
    // selection helper handles invalid child typename
    const { selections, ...rest } = buildSelections(props)(child)

    const field = createField({
      fieldname,
      arguments: inputs.args,
      selections,
    })

    return {
      ...rest,
      variables: inputs.vars,
      selections: [field],
    }
  }

  // noop
  throw Error
}
