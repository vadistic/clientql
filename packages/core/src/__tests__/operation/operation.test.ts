import { COMPLEX_TYPEDEFS } from '@graphql-clientgen/testing'
import { print } from 'graphql'
import { FragmentType, getCoreProps } from '../../config'
import { buildOperation } from '../../operation'
import { complexProps, prismaProps } from '../fixture'

describe('operation', () => {
  it('prisma > non nested single query', () => {
    const op = buildOperation(prismaProps)(['user'])!

    expect(print(op)).toMatchSnapshot()
  })

  it('prisma > non nested multi query', () => {
    const op = buildOperation(prismaProps)(['users'])!

    expect(print(op)).toMatchSnapshot()
  })

  it('prisma > nested query', () => {
    const op = buildOperation(prismaProps)(['user', 'posts'])!

    expect(print(op)).toMatchSnapshot()
  })

  it('complex > interface target', () => {
    const op = buildOperation(complexProps)(['findEventsAtVenue'])!

    expect(print(op)).toMatchSnapshot()
  })

  it('complex > deep fragments', () => {
    const props = getCoreProps(COMPLEX_TYPEDEFS, {
      useFragments: FragmentType.DEEP,
    })

    const op = buildOperation(props)(['findEventsAtVenue'])!

    expect(print(op)).toMatchSnapshot()
  })
})
