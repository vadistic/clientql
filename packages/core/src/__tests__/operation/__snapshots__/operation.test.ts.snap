// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation complex > deep fragments 1`] = `
"query TODO($venueId: ID!) {
  findEventsAtVenue(venueId: $venueId) {
    ...EventDeep
  }
}

fragment EventDeep on Event {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueFlat
  }
  minAgeRestriction
  ... on Concert {
    ...ConcertDeep
  }
  ... on Festival {
    ...FestivalDeep
  }
  ... on Conference {
    ...ConferenceDeep
  }
}

fragment VenueFlat on Venue {
  id
  name
  address
  maxOccupancy
}

fragment ConcertDeep on Concert {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueFlat
  }
  previousVenues {
    ...VenueFlat
  }
  minAgeRestriction
  performingBand {
    ...PerformerFlat
  }
}

fragment PerformerFlat on Performer {
  id
  name
  contact
  speciality
  fee
}

fragment FestivalDeep on Festival {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueFlat
  }
  minAgeRestriction
  performers {
    ...PerformerFlat
  }
}

fragment ConferenceDeep on Conference {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueFlat
  }
  minAgeRestriction
  speakers {
    ...SpeakerFlat
  }
  workshops
}

fragment SpeakerFlat on Speaker {
  id
  name
  contact
}
"
`;

exports[`operation complex > interface target 1`] = `
"query TODO($venueId: ID!) {
  findEventsAtVenue(venueId: $venueId) {
    ...EventFlat
    venue {
      ...VenueFlat
    }
    ... on Concert {
      ...ConcertFlat
      venue {
        ...VenueFlat
      }
      previousVenues {
        ...VenueFlat
      }
      performingBand {
        ...PerformerFlat
      }
    }
    ... on Festival {
      ...FestivalFlat
      venue {
        ...VenueFlat
      }
      performers {
        ...PerformerFlat
      }
    }
    ... on Conference {
      ...ConferenceFlat
      venue {
        ...VenueFlat
      }
      speakers {
        ...SpeakerFlat
      }
    }
  }
}

fragment EventFlat on Event {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment VenueFlat on Venue {
  id
  name
  address
  maxOccupancy
}

fragment ConcertFlat on Concert {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment PerformerFlat on Performer {
  id
  name
  contact
  speciality
  fee
}

fragment FestivalFlat on Festival {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment ConferenceFlat on Conference {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
  workshops
}

fragment SpeakerFlat on Speaker {
  id
  name
  contact
}
"
`;

exports[`operation prisma > nested query 1`] = `
"query TODO($where: UserWhereUniqueInput!, $where1: PostWhereInput, $orderBy1: PostOrderByInput, $skip1: Int, $after1: String, $before1: String, $first1: Int, $last1: Int) {
  user(where: $where) {
    posts(where: $where1, orderBy: $orderBy1, skip: $skip1, after: $after1, before: $before1, first: $first1, last: $last1) {
      ...PostFlat
      author {
        ...UserFlat
      }
    }
  }
}

fragment PostFlat on Post {
  id
  content
  createdAt
  published
  title
  updatedAt
}

fragment UserFlat on User {
  id
  nationality
  email
  name
}
"
`;

exports[`operation prisma > non nested multi query 1`] = `
"query TODO($where: UserWhereInput, $orderBy: UserOrderByInput, $skip: Int, $after: String, $before: String, $first: Int, $last: Int) {
  users(where: $where, orderBy: $orderBy, skip: $skip, after: $after, before: $before, first: $first, last: $last) {
    ...UserFlat
    posts {
      ...PostFlat
    }
  }
}

fragment UserFlat on User {
  id
  nationality
  email
  name
}

fragment PostFlat on Post {
  id
  content
  createdAt
  published
  title
  updatedAt
}
"
`;

exports[`operation prisma > non nested single query 1`] = `
"query TODO($where: UserWhereUniqueInput!) {
  user(where: $where) {
    ...UserFlat
    posts {
      ...PostFlat
    }
  }
}

fragment UserFlat on User {
  id
  nationality
  email
  name
}

fragment PostFlat on Post {
  id
  content
  createdAt
  published
  title
  updatedAt
}
"
`;
